### ISSUE TITLE
possible Magic Cookie regression bug

### ISSUE THREAD

**Issue by mjbludwig** [cea-hpc/modules#377]
Hi Xavier,

Forgive me if this is was already discussed prior, but I did not see a previous issue on this. 

I believe this bug was introduced around v4.2.0. I am using v4.5.2 (upstream Oracle Linux 8 rpm). I have a legacy cluster with module files that were generated as far back as v3.2.10 (at least) and the issue is with the Magic Cookie syntax that was acceptable back then.

Almost all of this clusters module files were generated with the Magic Cookie syntax `#%Modules`, the "s" is the issue. Back then there was no incorporation of appended env module version numbers and my guess is there was also no format checking. This bug results in module files without a Magic Cookie like `#%Module` or `#%Module1.0`, etc. being considered invalid and not listing in `module avail` but also not being able to be loaded. 

Unfortunately, I think this will also bite other clusters I admin in the near future as we continue to fight the never ending battle on moving to and keeping current distro releases (almost all Red Hat based, RIP CentOS...).

### To Reproduce

```console
[root@myserver /home/software/modulefiles]# cat hdf5/1.8.18 
#%Modules

conflict	 hdf5 
prepend-path	 PATH /home/software/hdf5/1.8.18/bin 
prepend-path	 LD_LIBRARY_PATH /home/software/hdf5/1.8.18/lib 
prepend-path	 CPATH /home/software/hdf5/1.8.18/include 
prepend-path	 LIBRARY_PATH /home/software/hdf5/1.8.18/lib 

[root@myserver /home/software/modulefiles]# module add hdf5/1.8.18

[root@fs11 ~]# module add hdf5/1.8.18
Module ERROR: Modulefile requires at least Modules version s
  In '/home/software/modulefiles/hdf5/1.8.18'
  Please contact <root@localhost>
```

Fwiw, this logic is clearly in modulecmd.tcl, maybe in the readModuleContent function (line ~6800 in v4.5.2)? The following example is my double check.

```console
[root@myserver /home/software/modulefiles]# cat hdf5/1.8.18 
#%Modulefoobar

conflict	 hdf5 
prepend-path	 PATH /home/software/hdf5/1.8.18/bin 
prepend-path	 LD_LIBRARY_PATH /home/software/hdf5/1.8.18/lib 
prepend-path	 CPATH /home/software/hdf5/1.8.18/include 
prepend-path	 LIBRARY_PATH /home/software/hdf5/1.8.18/lib 

[root@fs11 ~]# module add hdf5/1.8.18
Module ERROR: Modulefile requires at least Modules version foobar
  In '/home/software/modulefiles/hdf5/1.8.18'
  Please contact <root@localhost>

```

### Expected behavior

Ideally, being able to account for both `#%Module`  and `#%Modules` with or without appended version numbers.

I know this is pretty legacy but so is the nature of HPC in Academia.... 

### Error and debugging information

```console
$ module --version
Modules Release 4.5.2 (2020-07-30)


$ module config --dump-state
Modules Release 4.5.2 (2020-07-30)

- Config. name ---------.- Value (set by if default overridden) ---------------
advanced_version_spec     0
auto_handling             0
avail_indepth             1
avail_report_dir_sym      1
avail_report_mfile_sym    1
collection_pin_version    0
collection_target         <undef>
color                     never
colors                    hi=1:db=2:se=2:er=91:wa=93:me=95:in=94:mp=1;94:di=94:al=96:sy=95:de=4:cm=92
contact                   root@localhost
extended_default          0
extra_siteconfig          <undef>
home                      /usr/share/Modules (env-var)
icase                     never
ignored_dirs              CVS RCS SCCS .svn .git .SYNC .sos
implicit_default          1
locked_configs            
ml                        1
pager                     /usr/bin/less -eFKRX
rcfile                    <undef>
run_quarantine            LD_LIBRARY_PATH LD_PRELOAD (env-var)
search_match              starts_with
set_shell_startup         0
silent_shell_debug        <undef>
siteconfig                /etc/environment-modules/siteconfig.tcl
tcl_ext_lib               /usr/lib64/libtclenvmodules.so
term_background           dark
unload_match_order        returnlast
verbosity                 normal
wa_277                    0

- State name -----------.- Value ----------------------------------------------
cmdline                   /usr/share/Modules/libexec/modulecmd.tcl bash config --dump-state
extra_siteconfig_loaded   0
force                     0
is_stderr_tty             0
is_win                    0
machine                   x86_64
os                        Linux 5.4.17-2036.102.0.2.el8uek.x86_64
pager_started             0
paginate                  1
path_separator            :
rc_loaded                 <undef>
siteconfig_loaded         1
shell                     bash
shelltype                 sh
subcmd                    config
subcmd_args               --dump-state
tcl_ext_lib_loaded        1
tcl_version               8.6.8
term_columns              94

- Env. variable --------.- Value ----------------------------------------------
BASH_FUNC__module_raw%%   () {  unset _mlshdbg;
 if [ "${MODULES_SILENT_SHELL_DEBUG:-0}" = '1' ]; then
 case "$-" in 
 *v*x*)
 set +vx;
 _mlshdbg='vx'
 ;;
 *v*)
 set +v;
 _mlshdbg='v'
 ;;
 *x*)
 set +x;
 _mlshdbg='x'
 ;;
 *)
 _mlshdbg=''
 ;;
 esac;
 fi;
 unset _mlre _mlIFS;
 if [ -n "${IFS+x}" ]; then
 _mlIFS=$IFS;
 fi;
 IFS=' ';
 for _mlv in ${MODULES_RUN_QUARANTINE:-};
 do
 if [ "${_mlv}" = "${_mlv##*[!A-Za-z0-9_]}" -a "${_mlv}" = "${_mlv#[0-9]}" ]; then
 if [ -n "`eval 'echo ${'$_mlv'+x}'`" ]; then
 _mlre="${_mlre:-}${_mlv}_modquar='`eval 'echo ${'$_mlv'}'`' ";
 fi;
 _mlrv="MODULES_RUNENV_${_mlv}";
 _mlre="${_mlre:-}${_mlv}='`eval 'echo ${'$_mlrv':-}'`' ";
 fi;
 done;
 if [ -n "${_mlre:-}" ]; then
 eval `eval ${_mlre} /usr/bin/tclsh /usr/share/Modules/libexec/modulecmd.tcl bash '"$@"'`;
 else
 eval `/usr/bin/tclsh /usr/share/Modules/libexec/modulecmd.tcl bash "$@"`;
 fi;
 _mlstatus=$?;
 if [ -n "${_mlIFS+x}" ]; then
 IFS=$_mlIFS;
 else
 unset IFS;
 fi;
 unset _mlre _mlv _mlrv _mlIFS;
 if [ -n "${_mlshdbg:-}" ]; then
 set -$_mlshdbg;
 fi;
 unset _mlshdbg;
 return $_mlstatus
}
BASH_FUNC_module%%        () {  _module_raw "$@" 2>&1
}
LOADEDMODULES             
MODULEPATH                /usr/share/Modules/modulefiles:/etc/modulefiles:/usr/share/modulefiles:/home/software/modulefiles
MODULEPATH_modshare       /home/software/modulefiles:1:/usr/share/Modules/modulefiles:1:/etc/modulefiles:1:/usr/share/modulefiles:1
MODULESHOME               /usr/share/Modules
MODULES_CMD               /usr/share/Modules/libexec/modulecmd.tcl
MODULES_RUN_QUARANTINE    LD_LIBRARY_PATH LD_PRELOAD
mludwig@sait:~$ 

```

**Comment by xdelaruelle** [comment]
Hi Morgan,

Thanks a lot for this detailed report.

[Since Modules 4.2](https://modules.readthedocs.io/en/latest/MIGRATING.html#express-modules-compatibility-of-modulefile-with-versioned-magic-cookie), version set after `#%Module` string is checked to determine if modulefile is compatible with current module command version.

On version 3 of Modules, characters after the `#%Module` were not checked at all. However only a version number were expected after this string, even on this version.

I would suggest to update the existing modulefiles to remove the extra `s`.

Best that could be done is to add a new option to disable the version check on the magic cookie. This could be added in a new feature release.

**Comment by xdelaruelle** [comment]
A new configuration option will be available in Modules 4.7: [`mcookie_version_check`](https://modules.readthedocs.io/en/latest/module.html#mconfig-mcookie_version_check). When this option is set to `0`, the version set in magic cookie header is not checked.

**Comment by mjbludwig** [comment]
Hi Xavier,

Thanks for the quick response and subsequent pull request! I wish I had the time to have offered a pull request myself!

This is extremely helpful.
